test/tabulated_tests.jl

```diff
@@ -0,0 +1,225 @@
+
+
+"""
+    UnitTest(name, filename; meta)
+
+A unit test, given:
+ - `name::String` the name of the unit test
+ - `filename::String` the filename of the unit test
+ - `meta::Any` meta information for the test
+"""
+mutable struct UnitTest
+    name::String
+    filename::String
+    elapsed::Float64
+    meta::Any
+end
+UnitTest(name, filename; meta = nothing) = UnitTest(name, filename, 0.0, meta)
+
+"""
+    validate_tests(unit_tests::Vector{UnitTest}; test_path)
+
+Given:
+ - `unit_tests` a vector of `UnitTest`s
+ - `test_path` the path to the test directory (for checking that files exist).
+               Typically this should be `test_path = @__DIR__`.
+
+Returns `err::Symbol` indicating the validation results:
+
+ - `:duplicate_file` duplicate files found (and info statements are printed)
+ - `:non_existent_file` found non-existent files (and info statements are printed)
+ - `:pass` passes
+
+Checking for non-existent files can help
+prevent situations where a user adds a new
+unit test to the end of their test suite,
+but misspells the name. Instead of finding
+out at the end of the test suite, users can
+fail immediately due to a non-existent file.
+
+Checking for duplicate files can be helpful
+by avoiding unexpected duplicate work.
+"""
+function validate_tests(unit_tests::Vector{UnitTest}; test_path)
+    # Test uniqueness of included files
+    err = :pass
+    filenames = map(x -> x.filename, unit_tests)
+    if !allunique(filenames) # let's not do things more than once
+        for (key, val) in countmap(filenames)
+            val > 1 || continue
+            @info "Duplicate file found: $key, ($val times)"
+            err = :duplicate_file
+        end
+    end
+    # Test that files exist
+    if !all(map(x -> isfile(x.filename), unit_tests))
+        for filename in filenames
+            rfile = joinpath(test_path, filename)
+            isfile(rfile) && continue
+            @warn "Filename: $rfile does not exist"
+            err = :non_existent_file
+        end
+    end
+    return err
+end
+
+
+"""
+    tabulate_tests(
+        unit_tests::Vector{UnitTest};
+        include_timings::Bool = true,
+        time_format::Symbol = :second, # one of (:second, :compoundperiod)
+    )
+
+ - `include_timings::Bool` indicates whether or not to include the timings/percent columns
+ - `time_format::Symbol` specify the time format. Valid values include [:second (default), :compoundperiod].
+                         If we do not match either of these, then we warn & default to `:second`.
+
+Tabulate the given unit tests. The `include_timings` kwarg
+extends the table to include timings of the tests.
+"""
+function tabulate_tests(
+    unit_tests::Vector{UnitTest};
+    include_timings::Bool = true,
+    time_format::Symbol = :second,
+)
+    title =
+        include_timings ? "Tests results" :
+        "Running the following unit tests..."
+    if include_timings
+        sort!(unit_tests; by = x -> x.elapsed)
```

Comment:
This sorts time by ascending value, I think descending would be better for quickly checking long-running tests.

Reply:
Sure, I don't mind whether it's ascending/descending

Reply:
Ah, I forgot to reverse the table, I’ll follow up on this